### PR TITLE
added functionality to allow parent child relationships in elasticsearch

### LINKED ISSUE #26: Support Elasticsearch _parent field
https://www.elastic.co/guide/en/elasticsearch/reference/5.5/mapping-parent-field.html

### PR DESCRIPTION
**Pull request by JamesHollyer**
Added parent child relationships and a few debug help features. Will close #26

### CODE REVIEW

File: configParser.js [hunk 1]
```diff
@@ -59,6 +59,17 @@ if (mongoOpts.ssl) {
   mongoOpts.sslCert = fs.readFileSync(mongoOpts.sslCert, 'utf8');
 }
 
+//warn user if their collection names are too large to help debug errors in configuration
+let collectionNameTooLarge = false;
+CONFIG.mongo.collections.forEach((collection) => {
+  if(collection.length > 110){
```
Comment (badavis) [inline configParser.js]: Add a comment about the limit being 120 bytes and a link to this doc https://docs.mongodb.com/manual/reference/limits/#Restriction-on-Collection-Names

Comment (badavis) [reply configParser.js]: Also you could do something like this instead to be more explicit
```
const COLL_NAME_MAX_LENGTH;
CONFIG.mongo.collections.forEach((collection) => {
  if(collection.length > COLL_NAME_MAX_LENGTH){
    console.log(`WARNING: The collection ${collection} is ${COLL_NAME_MAX_LENGTH - collection.length} characters too long.  Mongo does not allow names to be larger that 120 bytes.`)
  }
});```

File: elasticManager.js [hunk 2]
```diff
@@ -42,16 +42,31 @@ class ElasticManager {
...
 
+    let parentId = this.getParentId(esReadyDoc, this.mappings[changeStreamObj.ns.coll].type);
+
     this.bulkOp.push({
         index:  {
           _index: this.mappings[changeStreamObj.ns.coll].index,
           _type: this.mappings[changeStreamObj.ns.coll].type,
-          _id: esId
+          _id: esId,
+          _parent:parentId
```
Comment (badavis) [inline elasticManager.js]: parentId being null doesn't cause any errors or unintended behavior right?

File: configParser.js (hunk 1 above)
Comment (JamesHollyer) [reply configParser.js]: good suggestions.  Done

File: elasticManager.js (hunk 2 above)
Comment (JamesHollyer) [reply elasticManager.js]: Nope.  In fact it doesn't even send across the network to ES.

File: configParser.js [hunk 3]
```diff
@@ -59,6 +59,15 @@ if (mongoOpts.ssl) {
   mongoOpts.sslCert = fs.readFileSync(mongoOpts.sslCert, 'utf8');
 }
 
+//warn user if their collection names are too large to help debug errors in configuration
+let collectionNameTooLarge = false;
```
Comment (badavis) [inline configParser.js]: this is an unused variable now

File: configParser.js [hunk 4]
```diff
@@ -59,6 +59,14 @@ if (mongoOpts.ssl) {
   mongoOpts.sslCert = fs.readFileSync(mongoOpts.sslCert, 'utf8');
 }
 
+//warn user if their collection names are too large to help debug errors in configuration
+const COLL_NAME_MAX_LENGTH = 110;
+CONFIG.mongo.collections.forEach((collection) => {
+  if(collection.length > COLL_NAME_MAX_LENGTH){
+    console.log(`WARNING: The collection ${collection} is ${collection.length - COLL_NAME_MAX_LENGTH} characters too large.  Mongo does not allow names to be larger that 120 bytes.(https://docs.mongodb.com/manual/reference/limits/#Restriction-on-Collection-Names)`);
```
Comment (dfinlay) [inline configParser.js]: The 110 character limit is intriguing. It appears the [actual limit is 120 bytes](https://docs.mongodb.com/manual/reference/limits/#Restriction-on-Collection-Names) for a collection namespace, database name and `.` delimiter inclusive.

Perhaps if we're going to go through the effort of preventing people from foot gunning, we should be precise:

```
const NAMESPACE_MAX_LENGTH = 120;

CONFIG.mongo.collections.forEach((collection) => {
  let ns_length = db.length + '.'.length + collection.length;
  if(ns_length > NAMESPACE_MAX_LENGTH){
    console.log(`WARNING: The collection ${collection}'s namespace is ${ns_length - NAMESPACE_MAX_LENGTH} characters too large.  Mongo does not allow names to be larger that 120 bytes.(https://docs.mongodb.com/manual/reference/limits/#Restriction-on-Collection-Names)`);
  }
});

```

Comment (dfinlay) [reply configParser.js]: Oh, you already had the 120 byte limit in there in the error.

Comment (JamesHollyer) [reply configParser.js]: when we tested creating a collection with a long name it maxed at 116 characters.  When we tried 117 it gave an error that it was too long.  I don't understand how that correlates to 120 bytes but we decided that just checking at 110 was reasonable.  It is only a warning to help debugging.  We will accept any length.

Comment (dfinlay) [reply configParser.js]: Sounds good to me, my suggestion was mostly in jest. Primarily an outlet for internal musings regarding the seemingly arbitrary MongoDB namespace max length. Maybe they are using UTF-16, maybe it's blood magic, maybe it sounded good on paper.

**Pull request merged**